test/core/test_task_launcher.py

```diff
@@ -74,20 +87,31 @@ def test_create_launch_expire_assignments(self):
         for assignment in launcher.assignments:
             self.assertEqual(assignment.get_status(), AssignmentState.CREATED)
 
-        launcher.launch_units("dummy-url:3000")
-
-        for unit in launcher.units:
-            self.assertEqual(unit.get_db_status(), AssignmentState.LAUNCHED)
-        for assignment in launcher.assignments:
-            self.assertEqual(assignment.get_status(), AssignmentState.LAUNCHED)
-
         launcher.expire_units()
 
         for unit in launcher.units:
             self.assertEqual(unit.get_db_status(), AssignmentState.EXPIRED)
         for assignment in launcher.assignments:
             self.assertEqual(assignment.get_status(), AssignmentState.EXPIRED)
 
+    def test_launch_assignments(self):
+        """Initialize a launcher on a task run, then create the assignments"""
+        mock_data_array = self.get_mock_assignment_data_array()
+        launcher = TaskLauncher(
+            self.db, self.task_run, mock_data_array, max_num_concurrent_units=1
+        )
+        launcher.launched_units = LimitedDict(launcher.max_num_concurrent_units)
+        launcher.create_assignments()
+        launcher.launch_units("dummy-url:3000")
+
+        for unit in launcher.units:
```

Comment:
This should be looping until all of the units are in the `COMPLETED` state, asserting that it's taken less than some amount of time to get there.

```
while not fully complete
  for all units
    if unit is launched, set it to complete
  sleep
  assert not over time limit
```

Reply:
I really missed the last line here. so sorry for that. But glad to know why.